fgpyo/io/__init__.py

```diff
@@ -0,0 +1,198 @@
+    # you find the first extent path, and check that it is a directory and is writeable.
+    elif path.is_dir() & parent_must_exist is False:
+        for directory in list(path.parents):
+            try:
+                assert (
+                    directory.is_dir() & directory.exists() & os.access(directory, os.W_OK)
+                ), f"File does not have a writeable parent directory: {path}"
+                return directory
+            except AssertionError:
+                pass
+        raise AssertionError(f"No parent directories of the path provided are writeable: {path}")
+
+
+def to_reader(path: Path) -> Union[io.TextIOWrapper, TextIO, IO[Any]]:
+    """Opens a Path for reading.
+
+    Args:
+        path: Path to read from
+
+    Example:
+    >>> reader = io.to_reader(path = Path("reader.txt"))
+    >>> reader.readlines()
+    >>> reader.close()
+    """
+    if path.suffix in COMPRESSED_FILE_EXTENSIONS:
+        return io.TextIOWrapper(cast(IO[bytes], gzip.open(path, mode="rb")), encoding="utf-8")
+    else:
+        return path.open(mode="r")
+
+
+def to_writer(path: Path) -> Union[IO[Any], io.TextIOWrapper]:
+    """Opens a Path for reading and based on extension uses open() or gzip.open()
+
+    Args:
+        path: Path to write to
+
+    Example:
+    >>> writer = io.to_writer(path = Path("writer.txt"))
+    >>> writer.write(f'{something}\n')
+    >>> writer.close()
+    """
+    if path.suffix in COMPRESSED_FILE_EXTENSIONS:
+        return io.TextIOWrapper(cast(IO[bytes], gzip.open(path, mode="wb")), encoding="utf-8")
+    else:
+        return path.open(mode="w")
+
+
+def read_lines(path: Path, strip: bool = True) -> Union[Iterator[str], Generator[str, None, None]]:
+    """Takes a path and reads each line into a generator, removing line terminators
+    along the way. Trailing and leading characters are stripped by default
+    however, there is an option to only strip trailing characters using the argument
+    `strip = False`.
+
+    Args:
+        path: Path to read from
+        strip: Boolean to strip lines or not
+
+    Example:
+    >>> read_back = io.read_lines(path)
+    """
+    with to_reader(path=path) as reader:
+        if strip:
+            for line in reader:
+                yield (f"{line!s}").strip()
+        else:
+            for line in reader:
+                yield f"{line!s}".rstrip("\r\n")
+
+
+def write_lines(path: Path, lines_to_write: Iterable[Any]) -> None:
+    """Writes a file with one line per item in provided iterable
+
+    Args:
+        path: Path to write to
+        lines_to_write: items to write to file
+
+    Example:
+    >>> lines: List[Any] = ["things to write", 100]
+    >>> path_to_write_to: Path = Path("file_to_write_to.txt")
+    >>> io.write_lines(path = path_to_write_to, lines_to_write = lines)
+    """
+    with to_writer(path=path) as writer:
+        for line in lines_to_write:
+            writer.write(f"{line}" + "\n")
```

Comment:
```suggestion
        for line in lines_to_write:
            writer.write(str(line))
            writer.write("\n")
```
Using an f-string and string concatenation are unnecessary.